### PR TITLE
16.0 werkzeug 222 moc

### PR DESCRIPTION
**Pull request by d-fence**
With the imminent release of Debian Bookworm which is packaging [werkzeug 2.2.2](https://packages.debian.org/bookworm/python3-werkzeug), this PR makes it possible to run odoo 16.0 with werkzeug >= 2.0.2.

Note:
This is an excerpt from this [testing branch](https://runbot.odoo.com/runbot/bundle/16-0-bookworm-moc-140562) that was made to test against the Jammy and Bookworm distributions.

### CODE REVIEW

File: addons/website/tests/test_lang_url.py [hunk 1]
```diff
@@ -118,32 +120,32 @@ def test_01_controller_redirect(self):
...
+        def assertUrlRedirect(url, expected_url, msg="", expected_query="", code=301):
             if not msg:
                 msg = 'Url <%s> differ from <%s>.' % (url, expected_url)
 
             r = self.url_open(url, head=True)
             self.assertEqual(r.status_code, code)
-            self.assertEqual(r.headers.get('Location'), expected_url, msg)
+            parsed_url = urlparse(r.headers.get('Location', ''))
+            self.assertEqual(parsed_url.query, expected_query)
+            self.assertEqual(parsed_url.path, expected_url, msg)
```
Comment (Julien00859) [inline addons/website/tests/test_lang_url.py:131]: So we don't have to change the function's signature and still parse both path and query
```py
url1 = urlparse(url)
url2 = urlparse(expected_url)
assertEqual(url1.path, url2.path, msg)
assertEqual(url1.query, url2.query, msg)

Comment (Julien00859) [reply addons/website/tests/test_lang_url.py:131]: Should restore the original code in the functions that use this `assertUrlRedirect` too

File: addons/web/tests/test_image.py [hunk 2]
```diff
@@ -103,7 +103,8 @@ def test_04_web_content_filename_secure(self):
         })
 
         res = self.url_open(f'/web/image/{att.id}')
-        self.assertEqual(res.headers['Content-Disposition'], 'inline; filename="foo-l\'eb _ a\\"!r\\".gif"; filename*=UTF-8\'\'f%C3%B4%E2%98%BAo-l%27%C3%A9b%20_%20a%22%21r%22.gif')
+        expected_ufilename = url_quote(att.name.replace('\n', '_').replace('\r', '_'))
+        self.assertEqual(res.headers['Content-Disposition'], f'inline; filename="foo-l\'eb _ a\\"!r\\".gif"; filename*=UTF-8\'\'{expected_ufilename}')
```
Comment (Julien00859) [inline addons/web/tests/test_image.py:107]: It's too hard to read :(
```suggestion
        self.assertEqual(res.headers['Content-Disposition'], r"""inline; filename="foo-l'eb _ a\"!r\".gif"; filename*=UTF-8''""" + expected_ufilename)
```

File: odoo/service/server.py [hunk 3]
```diff
@@ -138,6 +138,14 @@ def make_environ(self):
...
         return environ
 
+    def send_header(self, keyword, value):
+        # Prevent `WSGIRequestHandler` from sending the connection close header (compatibility with werkzeug >= 2.1.1 )
+        # since it is incompatible with websocket.
+        if self.headers.get('Upgrade') == 'websocket' and keyword == 'Connection' and value == 'close':
+            # Do not keep processing requests.
+            self.close_connection = True
+            return
+        super().send_header(keyword, value)
```
Comment (Julien00859) [inline odoo/service/server.py:148]: The commit says it is a "quick and dirty fix", what would be a nice and beautiful fix? cc @tsm-odoo

File: addons/website/tests/test_lang_url.py (hunk 1 above)
Comment (Julien00859) [reply addons/website/tests/test_lang_url.py:131]:
```py
        def assertUrlRedirect(url, expected_url, msg="", code=301):
            if not msg:
                msg = 'Url <%s> differ from <%s>.' % (url, expected_url)

            r = self.url_open(url, head=True)
            self.assertEqual(r.status_code, code)

            parsed_expected_url = urlparse(expected_url)
            parsed_location = urlparse(r.headers.get('Location', ''))

            self.assertEqual(parsed_location.path, parsed_expected_url.path, msg)
            self.assertEqual(parsed_location.query, parsed_expected_url.query, msg)
```

File: addons/website/tests/test_lang_url.py [hunk 4]
```diff
@@ -119,14 +121,15 @@ def test_01_controller_redirect(self):
...
             if not msg:
                 msg = 'Url <%s> differ from <%s>.' % (url, expected_url)
 
             r = self.url_open(url, head=True)
             self.assertEqual(r.status_code, code)
-            self.assertEqual(r.headers.get('Location'), expected_url, msg)
+            parsed_location = urlparse(r.headers.get('Location', ''))
+            parsed_expected_url = urlparse(expected_url)
+            self.assertEqual(parsed_location.query, parsed_expected_url.query)
+            self.assertEqual(parsed_location.path, parsed_expected_url.path, msg)
```
Comment (Julien00859) [inline addons/website/tests/test_lang_url.py:132]: Testing the path first would produce a better error message (e.g. "/web/login" != "/my/auth/user/route"), the custom message should be used for an unequal query too
```suggestion
            self.assertEqual(parsed_location.path, parsed_expected_url.path, msg)
            self.assertEqual(parsed_location.query, parsed_expected_url.query, msg)
```

**Review by Julien00859 — APPROVED** [review]
@robodoo r+

**Comment by d-fence** [comment]
@robodoo rebase-ff

**Review by tde-banana-odoo — COMMENTED** [review]
Don't know why I am pinged, so will say: coucou Moc :wave:  !

**Pull request closed without merging**